hack/generators/cache-stores-snapshot/main.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"bytes"
+	"os"
+	"reflect"
+	"text/template"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/store"
+)
+
+// This code generates `store.CacheStores.TakeSnapshot` method.
+
+const (
+	outputFile = "zz_generated_cache_stores_snapshot.go"
+
+	outputTemplate = `// Code generated by hack/generators/cache/snapshot.go; DO NOT EDIT.
+package store
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (c CacheStores) TakeSnapshot() (CacheStores, error) {
+	c.l.RLock()
+	defer c.l.RUnlock()
+	
+	// Create a fresh CacheStores instance to store the snapshot.
+	snapshot := NewCacheStores()
```

Comment:
This is not required to be guarded by a mutex, but on the other hand, it's not rather so heavy to spot differences that unnecessary computations are guarded